subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,775 @@
+{
+	int64_t remaining = APP_SLEEP_MSECS;
+	int64_t start_time = k_uptime_get();
+	struct shell_mqtt *ctx;
+	int rc;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	/* Listen to the port for 500 ms as defined by APP_SLEEP_MSECS */
+	while (remaining > 0) {
+		if (wait(remaining)) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+				goto process_err;
+			}
+		}
+
+		rc = mqtt_live(&ctx->client_ctx);
+		if (rc != 0 && rc != -EAGAIN) {
+			LOG_ERR("mqtt_live error: %d", rc);
+			goto process_err;
+		}
+
+		remaining = APP_SLEEP_MSECS + start_time - k_uptime_get();
+	}
+
+	/* Reschedule the work */
+	k_work_init_delayable(&ctx->mqtt_work, shell_mqtt_process_work);
+	(void)k_work_reschedule_for_queue(&sh_mqtt->sh_mqtt_workq, &sh_mqtt->mqtt_work,
+					  K_SECONDS(MQTT_PROCESS_INTERVAL));
+	return;
+
+process_err:
+	ctx.state = SHELL_MQTT_OP_ERROR;
+	shell_mqtt_handle_error(ctx);
+}
+
+static int shell_mqtt_subscribe_config(struct mqtt_client *const client)
+{
+	/* subscribe to config information */
+	/* clang-format off */
+	struct mqtt_topic subs_topic = {
+		.topic = {
+			.utf8 = sh_mqtt->sub_topic,
+			.size = strlen(sh_mqtt->sub_topic)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+	const struct mqtt_subscription_list subs_list = {
+		.list = &subs_topic,
+		.list_count = 1U,
+		.message_id = 1U
+	};
+	/* clang-format on */
+	int rc;
+
+	rc = mqtt_subscribe(client, &subs_list);
+	if (rc) {
+		LOG_ERR("Failed to subscribe to %s item, error %d", subs_topic.topic.utf8, rc);
+	}
+
+	return rc;
+}
+
+static void shell_mqtt_subscribe_work(struct k_work *work)
+{
+	struct shell_mqtt *ctx;
+	int rc;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	rc = shell_mqtt_subscribe_config(&ctx->client_ctx);
+	if (rc == 0) {
+		/* Change the work to process/keepalive */
+		LOG_DBG("Scheduling MQTT process & keepalive work");
+		k_work_init_delayable(&ctx->mqtt_work, shell_mqtt_process_work);
+		(void)k_work_reschedule_for_queue(&sh_mqtt->sh_mqtt_workq, &sh_mqtt->mqtt_work,
+						  K_SECONDS(MQTT_PROCESS_INTERVAL));
+	} else {
+		ctx.state = SHELL_MQTT_OP_ERROR;
+		shell_mqtt_handle_error(ctx);
+	}
+}
+
+static void shell_mqtt_connect_work(struct k_work *work)
+{
+	int rc;
+	struct shell_mqtt *ctx;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	/* Try to connect to mqtt */
+	rc = mqtt_connect(&ctx->client_ctx);
+	if (rc != 0) {
+		LOG_ERR("mqtt_connect error: %d", rc);
+		goto conn_fail;
+	}
+
+	/* Prepare port config */
+	prepare_fds(ctx);
+
+	/* Wait for mqtt's connack */
+	if (wait(APP_CONNECT_TIMEOUT_MS)) {
+		ctx->state = SHELL_MQTT_OP_CONNECTING;
+		rc = mqtt_input(&ctx->client_ctx);
+		if (rc != 0) {
+			LOG_ERR("mqtt_input error: %d", rc);
+			goto conn_fail;
+		}
+	}
+
+	/* No connack, fail */
+	if (ctx->state != SHELL_MQTT_OP_CONNECTED) {
+		goto conn_fail;
+	}
+
+	LOG_DBG("MQTT connected successfully");
+	ctx->state = SHELL_MQTT_OP_CONNECTED;
```

Comment:
Shouldn't this be already set by CONNACK handler?

Reply:
> The MQTT part is looking pretty good now, still needs some cleanup with regards to commented out code, but it seems to go into the right direction.

Thanks for all the reviews, update coming this weekend!